# USGS dataretrieval Python Package `get_samples()` Examples

This notebook provides examples of using the Python dataretrieval package to retrieve water quality sample data for United States Geological Survey (USGS) monitoring sites. The dataretrieval package provides a collection of functions to get data from the USGS Samples database and other online sources of hydrology and water quality data, including the United States Environmental Protection Agency (USEPA).

### Install the Package

Use the following code to install the package if it doesn't exist already within your Jupyter Python environment.

In [1]:
!pip install dataretrieval

Defaulting to user installation because normal site-packages is not writeable


Load the package so you can use it along with other packages used in this notebook.

In [2]:
from dataretrieval import waterdata
from IPython.display import display

### Basic Usage

The dataretrieval package has several functions that allow you to retrieve data from different web services. This examples uses the `get_samples()` function to retrieve water quality sample data for USGS monitoring sites from Samples. The following arguments are supported:

* **ssl_check** : boolean, optional
        Check the SSL certificate.
* **service** : string
        One of the available Samples services: "results", "locations", "activities",
        "projects", or "organizations". Defaults to "results".
* **profile** : string
        One of the available profiles associated with a service. Options for each
        service are:
        results - "fullphyschem", "basicphyschem",
                    "fullbio", "basicbio", "narrow",
                    "resultdetectionquantitationlimit",
                    "labsampleprep", "count"
        locations - "site", "count"
        activities - "sampact", "actmetric",
                        "actgroup", "count"
        projects - "project", "projectmonitoringlocationweight"
        organizations - "organization", "count"
* **activityMediaName** : string or list of strings, optional
        Name or code indicating environmental medium in which sample was taken.
        Check the `activityMediaName_lookup()` function in this module for all
        possible inputs.
        Example: "Water".
* **activityStartDateLower** : string, optional
        The start date if using a date range. Takes the format YYYY-MM-DD.
        The logic is inclusive, i.e. it will also return results that
        match the date. If left as None, will pull all data on or before
        activityStartDateUpper, if populated.
* **activityStartDateUpper** : string, optional
        The end date if using a date range. Takes the format YYYY-MM-DD.
        The logic is inclusive, i.e. it will also return results that
        match the date. If left as None, will pull all data after
        activityStartDateLower up to the most recent available results.
* **activityTypeCode** : string or list of strings, optional
        Text code that describes type of field activity performed.
        Example: "Sample-Routine, regular".
* **characteristicGroup** : string or list of strings, optional
        Characteristic group is a broad category of characteristics
        describing one or more results. Check the `characteristicGroup_lookup()`
        function in this module for all possible inputs.
        Example: "Organics, PFAS"
* **characteristic** : string or list of strings, optional
        Characteristic is a specific category describing one or more results.
        Check the `characteristic_lookup()` function in this module for all
        possible inputs.
        Example: "Suspended Sediment Discharge"
* **characteristicUserSupplied** : string or list of strings, optional
        A user supplied characteristic name describing one or more results.
* **boundingBox**: list of four floats, optional
        Filters on the the associated monitoring location's point location
        by checking if it is located within the specified geographic area. 
        The logic is inclusive, i.e. it will include locations that overlap
        with the edge of the bounding box. Values are separated by commas,
        expressed in decimal degrees, NAD83, and longitudes west of Greenwich
        are negative.
        The format is a string consisting of:
        - Western-most longitude
        - Southern-most latitude
        - Eastern-most longitude
        - Northern-most longitude 
        Example: [-92.8,44.2,-88.9,46.0]
* **countryFips** : string or list of strings, optional
        Example: "US" (United States)
* **stateFips** : string or list of strings, optional
        Check the `stateFips_lookup()` function in this module for all
        possible inputs.
        Example: "US:15" (United States: Hawaii)
* **countyFips** : string or list of strings, optional
        Check the `countyFips_lookup()` function in this module for all
        possible inputs.
        Example: "US:15:001" (United States: Hawaii, Hawaii County)
* **siteTypeCode** : string or list of strings, optional
        An abbreviation for a certain site type. Check the `siteType_lookup()`
        function in this module for all possible inputs.
        Example: "GW" (Groundwater site)
* **siteTypeName** : string or list of strings, optional
        A full name for a certain site type. Check the `siteType_lookup()`
        function in this module for all possible inputs.
        Example: "Well"
* **usgsPCode** : string or list of strings, optional
        5-digit number used in the US Geological Survey computerized
        data system, National Water Information System (NWIS), to
        uniquely identify a specific constituent. Check the 
        `characteristic_lookup()` function in this module for all possible
        inputs.
        Example: "00060" (Discharge, cubic feet per second)
* **hydrologicUnit** : string or list of strings, optional
        Max 12-digit number used to describe a hydrologic unit.
        Example: "070900020502"
* **monitoringLocationIdentifier** : string or list of strings, optional
        A monitoring location identifier has two parts: the agency code
        and the location number, separated by a dash (-).
        Example: "USGS-040851385"
* **organizationIdentifier** : string or list of strings, optional
        Designator used to uniquely identify a specific organization.
        Currently only accepting the organization "USGS".
* **pointLocationLatitude** : float, optional
        Latitude for a point/radius query (decimal degrees). Must be used
        with pointLocationLongitude and pointLocationWithinMiles.
* **pointLocationLongitude** : float, optional
        Longitude for a point/radius query (decimal degrees). Must be used
        with pointLocationLatitude and pointLocationWithinMiles.
* **pointLocationWithinMiles** : float, optional
        Radius for a point/radius query. Must be used with
        pointLocationLatitude and pointLocationLongitude
* **projectIdentifier** : string or list of strings, optional
        Designator used to uniquely identify a data collection project. Project
        identifiers are specific to an organization (e.g. USGS).
        Example: "ZH003QW03"
* **recordIdentifierUserSupplied** : string or list of strings, optional
        Internal AQS record identifier that returns 1 entry. Only available
        for the "results" service.

#### Example 1: Get all water quality sample data for a single monitoring site

In [3]:
siteID = 'USGS-10109000'
wq_data = waterdata.get_samples(monitoringLocationIdentifier=siteID)
print('Retrieved data for ' + str(len(wq_data[0])) + ' samples.')

Retrieved data for 2333 samples.


### Interpreting the Result

The result of calling the `get_samples()` function is an object that contains a Pandas data frame object and an associated metadata object. The Pandas data frame contains the water quality sample data for the requested site, and or observed variables and time frame.

Once you've got the data frame, there's several useful things you can do to explore the data.

Display the data frame as a table. The default data frame for this function is a  long, flat table, with a row for each observed variable at a given site and date/time.

In [4]:
display(wq_data[0])

Org_Identifier          Org_FormalName    Project_Identifier  \
0              USGS  U.S. Geological Survey              ["USGS"]   
1              USGS  U.S. Geological Survey              ["USGS"]   
2              USGS  U.S. Geological Survey              ["USGS"]   
3              USGS  U.S. Geological Survey  ["USGS","UT2110000"]   
4              USGS  U.S. Geological Survey  ["USGS","UT1710001"]   
...             ...                     ...                   ...   
2328           USGS  U.S. Geological Survey  ["USGS","UT2610000"]   
2329           USGS  U.S. Geological Survey  ["USGS","UT2610000"]   
2330           USGS  U.S. Geological Survey  ["USGS","UT2610000"]   
2331           USGS  U.S. Geological Survey  ["USGS","UT2610000"]   
2332           USGS  U.S. Geological Survey  ["USGS","UT2610000"]   

      Project_Name  Project_QAPPApproved  Project_QAPPApprovalAgency  \
0              NaN                   NaN                         NaN   
1              NaN                   NaN                         NaN   
2              NaN                   NaN                         NaN   
3              NaN                   NaN                         NaN   
4              NaN                   NaN                         NaN   
...            ...                   ...                         ...   
2328           NaN                   NaN                         NaN   
2329           NaN                   NaN                         NaN   
2330           NaN                   NaN                         NaN   
2331           NaN                   NaN                         NaN   
2332           NaN                   NaN                         NaN   

      ProjectAttachment_FileName  ProjectAttachment_FileType  \
0                            NaN                         NaN   
1                            NaN                         NaN   
2                            NaN                         NaN   
3                            NaN                         NaN   
4                            NaN                         NaN   
...                          ...                         ...   
2328                         NaN                         NaN   
2329                         NaN                         NaN   
2330                         NaN                         NaN   
2331                         NaN                         NaN   
2332                         NaN                         NaN   

     Location_Identifier                                Location_Name  ...  \
0          USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
1          USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
2          USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
3          USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
4          USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
...                  ...                                          ...  ...   
2328       USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
2329       USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
2330       USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
2331       USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   
2332       USGS-10109000  LOGAN RIVER ABOVE STATE DAM, NEAR LOGAN, UT  ...   

     ResultAttachment_FileName  ResultAttachment_FileType  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2328                       NaN                        NaN   
2329                       NaN                        NaN   
2330                       NaN                        Na

Show the data types of the columns in the resulting data frame.

In [5]:
print(wq_data[0].dtypes)

Org_Identifier                 object
Org_FormalName                 object
Project_Identifier             object
Project_Name                  float64
Project_QAPPApproved          float64
                               ...   
Result_CharacteristicGroup     object
Org_Type                       object
LastChangeDate                 object
USGSpcode                       int64
USGSSampleAquifer             float64
Length: 181, dtype: object


The other part of the result returned from the `get_samples()` function is a metadata object that contains information about the query that was executed to return the data. For example, you can access the URL that was assembled to retrieve the requested data from the USGS web service. The USGS web service responses contain a descriptive header that defines and can be helpful in interpreting the contents of the response.

In [6]:
print('The query URL used to retrieve the data from USGS Samples was: ' + wq_data[1].url)

The query URL used to retrieve the data from USGS Samples was: https://api.waterdata.usgs.gov/samples-data/results/fullphyschem?monitoringLocationIdentifier=USGS-10109000&mimeType=text%2Fcsv


### Additional Examples

#### Example 2: Get water quality sample data for multiple sites for a single parameter

In [7]:
site_ids = ['USGS-04024430', 'USGS-04024000']
parameter_code = '00065'
wq_multi_site = waterdata.get_samples(monitoringLocationIdentifier=site_ids, usgsPCode=parameter_code)
print('Retrieved data for ' + str(len(wq_multi_site[0])) + ' samples.')
display(wq_multi_site[0])

Retrieved data for 303 samples.


Org_Identifier          Org_FormalName  \
0             USGS  U.S. Geological Survey   
1             USGS  U.S. Geological Survey   
2             USGS  U.S. Geological Survey   
3             USGS  U.S. Geological Survey   
4             USGS  U.S. Geological Survey   
..             ...                     ...   
298           USGS  U.S. Geological Survey   
299           USGS  U.S. Geological Survey   
300           USGS  U.S. Geological Survey   
301           USGS  U.S. Geological Survey   
302           USGS  U.S. Geological Survey   

                                    Project_Identifier  Project_Name  \
0    ["Great Lakes Restoration Initiative","GR13NK0...           NaN   
1    ["Great Lakes Restoration Initiative","MN-00300"]           NaN   
2    ["Great Lakes Restoration Initiative","GR12NK0...           NaN   
3    ["Great Lakes Restoration Initiative","00GQ436...           NaN   
4    ["Great Lakes Restoration Initiative","MN-00300"]           NaN   
..                                                 ...           ...   
298   ["Great Lakes Restoration Initiative","NA00V18"]           NaN   
299   ["Great Lakes Restoration Initiative","NA00V18"]           NaN   
300   ["Great Lakes Restoration Initiative","NA00V18"]           NaN   
301   ["Great Lakes Restoration Initiative","NA00V18"]           NaN   
302   ["Great Lakes Restoration Initiative","NA00V18"]           NaN   

     Project_QAPPApproved  Project_QAPPApprovalAgency  \
0                     NaN                         NaN   
1                     NaN                         NaN   
2                     NaN                         NaN   
3                     NaN                         NaN   
4                     NaN                         NaN   
..                    ...                         ...   
298                   NaN                         NaN   
299                   NaN                         NaN   
300                   NaN                         NaN   
301                   NaN                         NaN   
302                   NaN                         NaN   

     ProjectAttachment_FileName  ProjectAttachment_FileType  \
0                           NaN                         NaN   
1                           NaN                         NaN   
2                           NaN                         NaN   
3                           NaN                         NaN   
4                           NaN                         NaN   
..                          ...                         ...   
298                         NaN                         NaN   
299                         NaN                         NaN   
300                         NaN                         NaN   
301                         NaN                         NaN   
302                         NaN                         NaN   

    Location_Identifier                   Location_Name  ...  \
0         USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
1         USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
2         USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
3         USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
4         USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
..                  ...                             ...  ...   
298       USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
299       USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
300       USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
301       USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   
302       USGS-04024000  ST. LOUIS RIVER AT SCANLON, MN  ...   

    ResultAttachment_FileName  ResultAttachment_FileType  \
0                         NaN                        NaN   
1                         NaN                        NaN   
2                         NaN                        NaN   
3                         NaN                        NaN   
4                         NaN                        NaN   
..             

#### Example 3: Retrieve water quality sample data for multiple sites, including a list of parameters, within a time period defined by start date until present

In [8]:
site_ids = ['USGS-04024430', 'USGS-04024000']
parameterCd = ['34247', '30234', '32104', '34220']
startDate = '2012-01-01'
wq_data2 = waterdata.get_samples(monitoringLocationIdentifier=site_ids, usgsPCode=parameterCd,
                           activityStartDateLower=startDate)
print('Retrieved data for ' + str(len(wq_multi_site[0])) + ' samples.')
display(wq_data2[0])


Retrieved data for 303 samples.


Org_Identifier          Org_FormalName  \
0             USGS  U.S. Geological Survey   
1             USGS  U.S. Geological Survey   
2             USGS  U.S. Geological Survey   
3             USGS  U.S. Geological Survey   
4             USGS  U.S. Geological Survey   
..             ...                     ...   
147           USGS  U.S. Geological Survey   
148           USGS  U.S. Geological Survey   
149           USGS  U.S. Geological Survey   
150           USGS  U.S. Geological Survey   
151           USGS  U.S. Geological Survey   

                                    Project_Identifier  Project_Name  \
0                                     ["USGS","F6X7H"]           NaN   
1    ["Great Lakes Restoration Initiative","00GQ414...           NaN   
2                                  ["USGS","GLRI TOX"]           NaN   
3    ["Great Lakes Restoration Initiative","GR12NK0...           NaN   
4    ["Great Lakes Restoration Initiative","MN-00300"]           NaN   
..                                                 ...           ...   
147  ["Great Lakes Restoration Initiative","MN-00300"]           NaN   
148  ["Great Lakes Restoration Initiative","GR14NK0...           NaN   
149                                ["USGS","GLRI TOX"]           NaN   
150  ["Great Lakes Restoration Initiative","GR12NK0...           NaN   
151  ["Great Lakes Restoration Initiative","GR12NK0...           NaN   

     Project_QAPPApproved  Project_QAPPApprovalAgency  \
0                     NaN                         NaN   
1                     NaN                         NaN   
2                     NaN                         NaN   
3                     NaN                         NaN   
4                     NaN                         NaN   
..                    ...                         ...   
147                   NaN                         NaN   
148                   NaN                         NaN   
149                   NaN                         NaN   
150                   NaN                         NaN   
151                   NaN                         NaN   

     ProjectAttachment_FileName  ProjectAttachment_FileType  \
0                           NaN                         NaN   
1                           NaN                         NaN   
2                           NaN                         NaN   
3                           NaN                         NaN   
4                           NaN                         NaN   
..                          ...                         ...   
147                         NaN                         NaN   
148                         NaN                         NaN   
149                         NaN                         NaN   
150                         NaN                         NaN   
151                         NaN                         NaN   

    Location_Identifier                          Location_Name  ...  \
0         USGS-04024430  NEMADJI RIVER NEAR SOUTH SUPERIOR, WI  ...   
1         USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
2         USGS-04024430  NEMADJI RIVER NEAR SOUTH SUPERIOR, WI  ...   
3         USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
4         USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
..                  ...                                    ...  ...   
147       USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
148       USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
149       USGS-04024430  NEMADJI RIVER NEAR SOUTH SUPERIOR, WI  ...   
150       USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   
151       USGS-04024000         ST. LOUIS RIVER AT SCANLON, MN  ...   

    ResultAttachment_FileName  ResultAttachment_FileType  \
0                         NaN                        NaN   
1                         NaN                        NaN   
2                         NaN                        NaN   
3                         NaN                      

#### Example 4: Retrieve water quality sample data for one site and convert to a wide format

Note that the USGS Samples database returns multiple parameters in a "long" format: each row in the resulting table represents a single observation of a single parameters. Furthermore, every observation has 181 fields of metadata. However, if you wanted to place your water quality data into a "wide" format, where each column represents a water quality parameter code, the code below details one solution.

In [9]:
siteID = 'USGS-10109000'
wq_data,_ = waterdata.get_samples(monitoringLocationIdentifier=siteID)
print('Retrieved data for ' + str(len(wq_data)) + ' sample results.')

wq_data["characteristic_unit"] = wq_data["Result_Characteristic"] + ", " + wq_data["Result_MeasureUnit"]
wq_data_wide = wq_data.pivot_table(index=['Location_Identifier', 'Activity_StartDate', 'Activity_StartTime'], columns="characteristic_unit", values="Result_Measure", aggfunc='first')
display(wq_data_wide)


Retrieved data for 2333 sample results.


characteristic_unit                                       Acidity, (H+), mg/L  \
Location_Identifier Activity_StartDate Activity_StartTime                       
USGS-10109000       1967-09-13         07:35:00                       0.00001   
                    1968-01-18         12:20:00                       0.00002   
                    1968-05-15         12:30:00                       0.00004   
                    1968-07-26         14:40:00                       0.00002   
                    1972-12-08         16:15:00                           NaN   
...                                                                       ...   
                    2025-06-17         11:31:00                           NaN   
                    2025-07-08         12:14:00                           NaN   
                    2025-08-06         15:07:00                           NaN   
                    2025-09-15         13:41:00                           NaN   
                    2025-11-18         08:55:00                           NaN   

characteristic_unit                                       Alkalinity, mg/L  \
Location_Identifier Activity_StartDate Activity_StartTime                    
USGS-10109000       1967-09-13         07:35:00                        187   
                    1968-01-18         12:20:00                        207   
                    1968-05-15         12:30:00                        149   
                    1968-07-26         14:40:00                        179   
                    1972-12-08         16:15:00                        NaN   
...                                                                    ...   
                    2025-06-17         11:31:00                        NaN   
                    2025-07-08         12:14:00                        NaN   
                    2025-08-06         15:07:00                        NaN   
                    2025-09-15         13:41:00                        NaN   
                    2025-11-18         08:55:00                        NaN   

characteristic_unit                                       Bicarbonate, mg/L  \
Location_Identifier Activity_StartDate Activity_StartTime                     
USGS-10109000       1967-09-13         07:35:00                         228   
                    1968-01-18         12:20:00                         252   
                    1968-05-15         12:30:00                         182   
                    1968-07-26         14:40:00                         218   
                    1972-12-08         16:15:00                         NaN   
...                                                                     ...   
                    2025-06-17         11:31:00                         NaN   
                    2025-07-08         12:14:00                         NaN   
                    2025-08-06         15:07:00                         NaN   
                    2025-09-15         13:41:00                         NaN   
                    2025-11-18         08:55:00                         NaN   

characteristic_unit                                       Calcium, mg/L  \
Location_Identifier Activity_StartDate Activity_StartTime                 
USGS-10109000       1967-09-13         07:35:00                    44.0   
                    1968-01-18         12:20:00                    52.0   
                    1968-05-15         12:30:00                    38.0   
                    1968-07-26         14:40:00                    53.0   
                    1972-12-08         16:15:00                     NaN   
...                                                                 ...   
                    2025-06-17         11:31:00                     NaN   
                    2025-07-08         12:14:00                     NaN   
                    2025-08-06         15:07:00                     NaN   
                    2025-09-15         13:41:00                     NaN   
   